File: web/js/renderers/starfieldRenderer.js
// Seeded PRNG — mulberry32
function mulberry32(seed) {
  return function () {
    seed |= 0;
    seed = (seed + 0x6d2b79f5) | 0;
    let t = Math.imul(seed ^ (seed >>> 15), 1 | seed);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// Draws a dark starry background with subtle twinkle animation.
export class StarfieldRenderer {
  constructor() {
    const rng = mulberry32(42);
    this.stars = [];
    this.starBrightness = [];
    for (let i = 0; i < 200; i++) {
      this.stars.push({ x: rng() * 2000, y: rng() * 2000 });
      this.starBrightness.push(0.3 + rng() * 0.7);
    }
    this.twinkleFrame = 0;
  }

  drawStarryBackground(ctx, x, y, w, h) {
    // Deep space gradient
    const bg = ctx.createLinearGradient(x + w / 2, y, x + w / 2, y + h);
    bg.addColorStop(0, 'rgb(8,8,40)');
    bg.addColorStop(1, 'rgb(2,2,18)');
    ctx.fillStyle = bg;
    ctx.fillRect(x, y, w, h);

    this.twinkleFrame++;

    for (let i = 0; i < this.stars.length; i++) {
      const sx = (this.stars[i].x % w) + x;
      const sy = (this.stars[i].y % h) + y;

      const twinkle =
        this.starBrightness[i] *
        (0.7 + 0.3 * Math.sin((this.twinkleFrame + i * 37) * 0.02));
      const alpha = twinkle;

      ctx.fillStyle = `rgba(255,255,240,${alpha})`;
      const size = this.starBrightness[i] > 0.8 ? 1.8 : 1.0;
      ctx.beginPath();
      ctx.arc(sx, sy, size, 0, Math.PI * 2);
      ctx.fill();
    }
  }
}

File: MoonSimulation/Renderers/StarfieldRenderer.cs
using SkiaSharp;

namespace MoonSimulation.Renderers;

/// <summary>
/// Draws a dark starry background with subtle twinkle animation.
/// Instance-based to own the star field state.
/// </summary>
public class StarfieldRenderer
{
    private readonly Random _rng = new(42); // Deterministic stars
    private SKPoint[]? _stars;
    private float[]? _starBrightness;
    private long _twinkleFrame;

    /// <summary>Current twinkle frame counter (read by other renderers for animation sync).</summary>
    public long TwinkleFrame => _twinkleFrame;

    /// <summary>Generated star positions (shared with sky scene for night stars).</summary>
    public SKPoint[]? Stars => _stars;

    /// <summary>Generated star brightness values.</summary>
    public float[]? StarBrightness => _starBrightness;

    /// <summary>
    /// Draws a deep-space starry background with twinkling stars.
    /// </summary>
    public void DrawStarryBackground(SKCanvas canvas, SKRect bounds)
    {
        // Deep space gradient
        using var bgPaint = new SKPaint();
        using var bgShader = SKShader.CreateLinearGradient(
            new SKPoint(bounds.MidX, bounds.Top),
            new SKPoint(bounds.MidX, bounds.Bottom),
            new SKColor[] { new(8, 8, 40), new(2, 2, 18) },
            null, SKShaderTileMode.Clamp);
        bgPaint.Shader = bgShader;
        canvas.DrawRect(bounds, bgPaint);

        // Generate stars once
        if (_stars == null || _stars.Length < 200)
        {
            _stars = new SKPoint[200];
            _starBrightness = new float[200];
            for (int i = 0; i < _stars.Length; i++)
            {
                _stars[i] = new SKPoint(
                    _rng.NextSingle() * 2000,
                    _rng.NextSingle() * 2000);
                _starBrightness[i] = 0.3f + _rng.NextSingle() * 0.7f;
            }
        }

        _twinkleFrame++;
        using var starPaint = new SKPaint { IsAntialias = true };

        for (int i = 0; i < _stars.Length; i++)
        {
            float x = _stars[i].X % bounds.Width + bounds.Left;
            float y = _stars[i].Y % bounds.Height + bounds.Top;

            // Gentle twinkle
            float twinkle = _starBrightness![i] *
                (0.7f + 0.3f * MathF.Sin((_twinkleFrame + i * 37) * 0.02f));
            byte alpha = (byte)(twinkle * 255);

            starPaint.Color = new SKColor(255, 255, 240, alpha);
            float size = _starBrightness[i] > 0.8f ? 1.8f : 1.0f;
            canvas.DrawCircle(x, y, size, starPaint);
        }
    }
}
